File: frontend/lib/features/tasks/repositories/task_repository.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:frontend/core/storage/hive_service.dart';
import 'package:frontend/features/tasks/models/task.dart';
import 'package:uuid/uuid.dart';

final taskRepositoryProvider = Provider<TaskRepository>((ref) {
  return TaskRepository();
});

class TaskRepository {
  Future<List<Task>> listTasks({String? status}) async {
    final box = HiveService.tasksBox;
    var tasks = box.values.toList();
    
    if (status != null) {
      tasks = tasks.where((t) => t.status.name == status).toList();
    }
    
    // Sort by due date (closest first), nulls at the end
    tasks.sort((a, b) {
      if (a.dueAt == null && b.dueAt == null) return 0;
      if (a.dueAt == null) return 1;
      if (b.dueAt == null) return -1;
      return a.dueAt!.compareTo(b.dueAt!);
    });
    
    return tasks;
  }

  Future<Task> getTask(String id) async {
    final box = HiveService.tasksBox;
    final task = box.get(id);
    if (task == null) {
      throw Exception('Task not found');
    }
    return task;
  }

  Future<Task> createTask({
    required String title,
    String? description,
    required String priority,
    DateTime? dueAt,
    String? customerId,
  }) async {
    final box = HiveService.tasksBox;
    final currentUser = HiveService.userBox.get('current_user');
    final now = DateTime.now();

    // Map string priority back to enum safely
    TaskPriority mappedPriority;
    switch (priority.toLowerCase()) {
      case 'low': mappedPriority = TaskPriority.low; break;
      case 'high': mappedPriority = TaskPriority.high; break;
      case 'medium': 
      default: mappedPriority = TaskPriority.medium; break;
    }

    final newTask = Task(
      id: const Uuid().v4(),
      userId: currentUser?.id ?? 'unknown_user',
      title: title,
      description: description,
      priority: mappedPriority,
      status: TaskStatus.pending,
      dueAt: dueAt,
      completedAt: null,
      createdAt: now,
      updatedAt: now,
      customerId: customerId,
    );

    await box.put(newTask.id, newTask);
    return newTask;
  }

  Future<Task> completeTask(String id) async {
    final box = HiveService.tasksBox;
    final existing = box.get(id);
    
    if (existing == null) {
      throw Exception('Task not found');
    }

    final updated = Task(
      id: existing.id,
      userId: existing.userId,
      title: existing.title,
      description: existing.description,
      priority: existing.priority,
      status: TaskStatus.completed,
      dueAt: existing.dueAt,
      completedAt: DateTime.now(),
      createdAt: existing.createdAt,
      updatedAt: DateTime.now(),
      customerId: existing.customerId,
    );

    await box.put(updated.id, updated);
    return updated;
  }

  Future<Task> updateTask(
    String id, {
    String? title,
    String? description,
    TaskPriority? priority,
    DateTime? dueAt,
    String? customerId,
  }) async {
    final box = HiveService.tasksBox;
    final existing = box.get(id);

    if (existing == null) {
      throw Exception('Task not found');
    }

    final updated = Task(
      id: existing.id,
      userId: existing.userId,
      title: title ?? existing.title,
      description: description ?? existing.description,
      priority: priority ?? existing.priority,
      status: existing.status,
      dueAt: dueAt ?? existing.dueAt,
      completedAt: existing.completedAt,
      createdAt: existing.createdAt,
      updatedAt: DateTime.now(),
      customerId: customerId ?? existing.customerId,
    );

    await box.put(updated.id, updated);
    return updated;
  }

  Future<void> deleteTask(String id) async {
    final box = HiveService.tasksBox;
    await box.delete(id);
  }
}

File: backend/app/infrastructure/repositories/task_repository.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.infrastructure.db.models.enums import TaskPriority, TaskStatus
from app.infrastructure.db.models.task import Task


class SqlAlchemyTaskRepository:
    def __init__(self, db: Session):
        self._db = db

    def get_by_id(self, task_id: uuid.UUID) -> Task | None:
        return self._db.get(Task, task_id)

    def list_by_user(self, user_id: uuid.UUID, *, status: TaskStatus | None = None) -> list[Task]:
        stmt = select(Task).where(Task.user_id == user_id)

        if status is not None:
            stmt = stmt.where(Task.status == status)

        stmt = stmt.order_by(Task.created_at.desc())

        return list(self._db.execute(stmt).scalars().all())

    def list_scoped(self, allowed_user_ids: list[uuid.UUID] | None = None, *, status: TaskStatus | None = None) -> list[Task]:
        stmt = select(Task)
        if allowed_user_ids is not None:
            stmt = stmt.where(Task.user_id.in_(allowed_user_ids))

        if status is not None:
            stmt = stmt.where(Task.status == status)

        stmt = stmt.order_by(Task.created_at.desc())

        return list(self._db.execute(stmt).scalars().all())

    def create_task(
        self,
        *,
        user_id: uuid.UUID,
        title: str,
        description: str | None,
        priority: TaskPriority,
        due_at: datetime | None,
    ) -> Task:
        task = Task(
            user_id=user_id,
            title=title,
            description=description,
            priority=priority,
            due_at=due_at,
        )
        self._db.add(task)
        self._db.flush()
        return task

    def complete_task(self, task: Task, completed_at: datetime) -> Task:
        task.status = TaskStatus.completed
        task.completed_at = completed_at
        self._db.flush()
        return task
